Support nested dictionaries

bpoldrack opened the pull request:
Closes #642

I think this closes #114, despite not yet addressing lists. WDYT, @aqw ?

aqw reviewed (changes requested):
I am /thrilled/ to see this feature.

Great PR. I left a lot of detailed feedback, most of which is exactly that: feedback and not blockers.

The primary thing I see that I would like to talk about is that keys of dicts are not returned in the key list. This creates some problems, but I would also like to understand what problems it solves.

Otherwise, huge +1 from me.

bpoldrack commented on onyo/lib/filters.py:
```diff
@@ -61,6 +61,8 @@ def match(self, asset: dict) -> bool:
                 asset[self.key] is None or asset[self.key] == ''):
             return True
         elif self.key not in asset.keys() or self.value == UNSET_VALUE:
+            # Note: This is weird.
```
Right, this shouldn't be in here, will fix that in a separate PR.
However, to clarify this:

- your second read is correct: The code amounts to an XOR and that's OK, it just reads weird and therefore isn't nice to maintain
- what it handles are two cases:
  - The key we are looking for is not in the asset and we are looking for a value other than `<unset>`. 
  - The key actually is present, but we are looking for `<unset>`.

aqw replied:
Ok. The comment referencing the previous `elif` statements made me think that the check was intentionally a complex intersection of the previous statements and current one.

Since the statements each `return`, there's no performance gained by using `elif`. In the PR you said you'll open, I propose changing these to `if` statements, so that one does not need to mentally load all four in order to understand the last one.

bpoldrack commented on onyo/lib/utils.py:
```diff
@@ -20,10 +20,125 @@
+    ['key', 'nested.key']
+
+    That implies that iteration considers the flattened view only. In the example above the key `nested`
+    (and its value - the dictionary) will not be yielded when using `wrapper.keys()`, `wrapper.values()`,
```
Not exactly. Such keys are still accessible, they are simply not yielded by iteration.

However, we'll talk about it ...

bpoldrack commented on onyo/lib/utils.py:
```diff
@@ -20,10 +20,125 @@
+                        effective_dict = effective_dict[parts[lvl]]
+                    except KeyError:
+                        # nested dict doesn't exist yet
+                        effective_dict[parts[lvl]] = dict()
```
> I like thinking about it as a convenience wrapper more than a data type.

Yes, that's what it is.

bpoldrack commented on onyo/lib/commands.py:
```diff
@@ -516,6 +517,13 @@ def onyo_get(inventory: Inventory,
                                             depth=depth,
                                             match=match)
+
+    # Problem: We may fill in stuff, that changes the structure. This approach did not consider nested dicts.
```
Your understanding is correct, but this entire comment is no longer relevant in removed in the next push.

bpoldrack commented on onyo/cli/tests/test_set.py:
```diff
@@ -62,8 +68,14 @@ def test_set(repo: OnyoRepo,
     for value in set_values:
-        assert value.replace("=", ": ") in Path.read_text(Path(asset))
+        k, v = value.split('=')
+        if '.' in k:
```
So, this is the one thing that is not addressed in the coming push. Not sure I even parse that right.
What should be looped over? You are literally commenting within a loop.

bpoldrack commented:
FTR: I don't intend to address the concerns code climate raises in this PR. I think this wrapper implementation will likely see a couple of changes over the coming weeks before it's time to actually simplify it a bit.

aqw commented on onyo/lib/command_utils.py:
```diff
@@ -62,29 +58,8 @@ def allowed_config_args(git_config_args: list[str]) -> bool:
     return True
 
 
-def fill_unset(assets: Generator[dict, None, None] | filter,
```
Nice to see this evaporate.

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
     for r in results:
         r['path'] = r['path'].relative_to(inventory.root)
 
+    # Note: Sorting is done before any further filtering/replacing. Therefore, one can sort by keys that aren't actually
```
Pure love for this feature.

This would be great information to add to the help text.

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
-    # filter output for `keys` only
-    results = [{k: v for k, v in r.items() if k in selected_keys} for r in results]
+    # Filter results for `selected_keys` first in order to not iterate over irrelevant parts of the assets in subsequent
+    # filters, replacements, etc.
```
Replacements makes sense, but filters, we don't filter any later. And if we did, this would be a bug here as onyo allows filtering (matching) on fields that aren't output.

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
+    results = [{k: r[k] if k in r else UNSET_VALUE
+                for k in selected_keys}
+               for r in results]
+    # Note: We now have a list of regular dicts. DotNotation would be resolved at this point by having keys with
```
> DotNotation would be resolved at this point ...

I don't quite parse this.

In general, I don't like the word "would" in comments. I really prefer stating what /is/ going on. We don't allow literal dots either. In effect, this comment increases my confusion rather than reducing it.

bpoldrack replied:
In the line above we create an output dictionary that actually has literal dots in its keys. So code following that never needs to consider the wrapper.
A result is not an asset. It's a collection of key-value pairs from an asset. And it's assembled in order to print it. At this point it's perfectly fine to not carry the original structure around.

I'll think about rephrasing.

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
+    # Note: We now have a list of regular dicts. DotNotation would be resolved at this point by having keys with
+    #       literal dots.
+
+    # Replace structures by an indication of type.
```
Nit: "by" -> "with"

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
+    #       literal dots.
+
+    # Replace structures by an indication of type.
+    # Instead of outputting `{}`, `[]` or even `{some: {other: value}}`, just print `<dict>`/`<list>`.
```
More of a comment rather than anything else, but I'm liking the idea that `<dict>` is the type, and would match `{}` or `{some: {other: value}}`. But `{}` would only match `{}` (i.e. an empty dict) and not ``{some: {other: value}}`. Same with lists, etc.

bpoldrack replied:
I'm a bit confused. May be just because where you put that comment.

This `{}` only matching empty dicts is what's implemented. But this piece of code is about printing the value in `get`'s output not about matching.

aqw replied:
Complete agreement here. My comment was indeed just a comment, and rather only loosely on-topic for where I chose to do so.

aqw commented on onyo/lib/commands.py:
```diff
@@ -513,22 +513,39 @@ def onyo_get(inventory: Inventory,
+
+    # Replace structures by an indication of type.
+    # Instead of outputting `{}`, `[]` or even `{some: {other: value}}`, just print `<dict>`/`<list>`.
+    # If information on content was wanted, the respective `keys` should be specified instead.
```
Nit: "was" -> "is" (only commenting because the confusion from the above "would" has continued to prime me. ;-)

aqw reviewed (approved):
Looks great! No blockers from my perspective. Some comments that would be nice to see fixed, but overall great. :-D

bpoldrack commented on onyo/lib/utils.py:
```diff
@@ -19,10 +19,127 @@
+    > list(wrapper.keys())
+    ['key', 'nested.key']
+
+    Iteration only considers the flattened view, and keys that contain dictionaries will not be yielded when using
```
Agree. list-keys is when this may become relevant.

bpoldrack commented on onyo/lib/tests/test_commands_unset.py:
```diff
@@ -299,3 +299,36 @@ def test_onyo_unset_asset_dir(inventory: Inventory) -> None:
+                 model=dict(name="MODEL"),
+                 serial="SERIAL2",
+                 directory=inventory.root)
+    test_pairs = dict(emptydict=dict(),
```
Good catch. Tested and fixed.

bpoldrack commented:
@aqw : Pushed again. Noteworthy: Added a test for get's output for `None`, `''` values as suggested. Turns out that did indeed not output `<unset>`. It now does. I adjusted a test for the opposite (in case of empty string) accordingly. While I find it debatable that `<unset>` matches keys that are not present as well as `None` and empty strings, this is what's implemented all along and out should match that idea. If, given that at least for an empty string the opposite was actually tested, you think we should re-consider, make it an issue please. I'd consider this out-of-scope here.

aqw commented on onyo/lib/commands.py:
```diff
@@ -483,6 +484,8 @@ def onyo_get(inventory: Inventory,
       `onyo.lib.consts.SORT_ASCENDING` and `onyo.lib.consts.SORT_DESCENDING`.
       If other values are specified an error is raised.
       Default: `{'path': SORT_ASCENDING}`.
+      Note, that this sorts the matching assets rather than the output.
```
Nit: The user only needs to know the feature.

Simply:

> One can sort by keys that are not in the output.

aqw commented on onyo/cli/get.py:
```diff
@@ -96,12 +100,14 @@
             Note, that if a **SORT-KEY** appears multiple times, the latest appearance will
-            overrule what ws specified before.
+            overrule what was specified before.
+            This sorts the matching assets rather than the output.
```
Nit: There's a tendency to over-explain here. But the user doesn't need to know /why/. They just need to know the feature.

Simply:

> One can sort by keys that are not in the output.

aqw commented on onyo/cli/get.py:
```diff
@@ -111,12 +117,14 @@
             This can be intermixed with ``-s/--sort-ascending``.
             Note, that if a **SORT-KEY** appears multiple times, the latest appearance will
             overrule what ws specified before.
+            This sorts the matching assets rather than the output.
```
Nit: The user only needs to know the feature.

Simply:

> One can sort by keys that are not in the output.

aqw commented on onyo/cli/get.py:
```diff
@@ -54,7 +56,9 @@
-            also be used. Special values supported are:
+            also be used. Dictionary subkeys can be addressed using a period
+            (e.g. ``model.name``, ``model.year``, etc.)
+            Special values supported are:
```
Nit: Due to the new text for sorting, I propose adding here:

> One can match keys that are not in the output.

aqw reviewed (approved):
I'm sold, and feel this is ready to merge.

I did leave comments on the help text. I am comfortable with those fixes being added to this PR, or making them myself outside of it.

Thanks for making this feature happen!

aqw commented:
> I think this closes #114, despite not yet addressing lists. WDYT, @aqw ?

I agree that this closes #114 (and does actually address initialization of lists).

bpoldrack commented:
Thanks for thoroughly reviewing this beast! :)
Merging.
